Spigot-Server-Patches/0446-Implement-Mob-Goal-API.patch

```diff
@@ -0,0 +1,446 @@
++        }
++    }
++
++    @Override
++    public void removeGoal(NamespacedKey key) {
++        for (Goal g : getAllGoals()) {
++            if (g.getKey().equals(key)) {
++                handle.removeGoal(((PaperVanillaGoal) g).getHandle());
++                break;
++            }
++        }
++    }
++
++    @Override
++    public boolean hasGoal(NamespacedKey key) {
++        for (Goal g : getAllGoals()) {
++            if (g.getKey().equals(key)) {
++                return true;
++            }
++        }
++        return false;
++    }
++
++    @Override
++    public Collection<Goal> getAllGoals() {
++        Set<Goal> goals = new HashSet<>();
++        for (PathfinderGoalSelector.PathfinderGoalSelectorItem item : getHandle().getAllTasks()) {
++            if (item.getGoal() instanceof PaperCustomGoal) {
++                goals.add(((PaperCustomGoal) item.getGoal()).getHandle());
++            } else {
++                goals.add(new PaperVanillaGoal(item.getGoal()));
++            }
++        }
++        return goals;
++    }
++
++    @Override
++    public Collection<Goal> getRunningGoals() {
++        Set<Goal> goals = new HashSet<>();
++        for (PathfinderGoalSelector.PathfinderGoalSelectorItem item : getHandle().getExecutingTasks()) {
++            if (item.getGoal() instanceof PaperCustomGoal) {
++                goals.add(((PaperCustomGoal) item.getGoal()).getHandle());
++            } else {
++                goals.add(new PaperVanillaGoal(item.getGoal()));
++            }
++        }
++        return goals;
++    }
++
++    public PathfinderGoalSelector getHandle() {
++        return handle;
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/ai/PaperVanillaGoal.java b/src/main/java/com/destroystokyo/paper/ai/PaperVanillaGoal.java
+new file mode 100644
+index 00000000..f7df35c3
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/ai/PaperVanillaGoal.java
+@@ -0,0 +1,87 @@
++package com.destroystokyo.paper.ai;
++
++import net.minecraft.server.PathfinderGoal;
++
++import org.bukkit.NamespacedKey;
++
++/**
++ * Wraps vanilla in api
++ */
++public class PaperVanillaGoal implements VanillaGoal {
++
++    private PathfinderGoal handle;
++    private NamespacedKey key;
++
++    public PaperVanillaGoal(PathfinderGoal handle) {
++        this.handle = handle;
++        this.key = NamespacedKey.minecraft(getUsableName(handle.getClass()));
++    }
++
++    public PathfinderGoal getHandle() {
++        return handle;
++    }
++
++    @Override
++    public boolean shouldActivate() {
++        return handle.shouldActivate();
++    }
++
++    @Override
++    public void start() {
++        handle.start();
++    }
++
++    @Override
++    public void stop() {
++        handle.onTaskReset();
++    }
++
++    @Override
++    public void tick() {
++        handle.tick();
++    }
++
++    @Override
++    public NamespacedKey getKey() {
++        return key;
++    }
++
++    static String getUsableName(Class<?> clazz) {
```

Comment:
before aikar finds this: I am totally gonna cache this

Reply:
already did, hence the note about 'avoiding the heavy logic' :P